README.md
# @malept/electron-installer-flatpak [![npm (scoped)](https://img.shields.io/npm/v/@malept/electron-installer-flatpak.svg)](https://www.npmjs.com/package/@malept/electron-installer-flatpak) [![Build Status](https://github.com/malept/electron-installer-flatpak/workflows/CI/badge.svg)](https://github.com/malept/electron-installer-flatpak/actions?query=workflow%3ACI) [![codecov](https://codecov.io/gh/malept/electron-installer-flatpak/branch/master/graph/badge.svg)](https://codecov.io/gh/malept/electron-installer-flatpak)


Create a flatpak for your Electron app. This is based off the
[electron-installer-debian](https://github.com/unindented/electron-installer-debian)
tool. Add flatpak support to an electron app using [electron-packager](https://github.com/electron-userland/electron-packager)
with minimal configuration.

## Requirements

This tool requires `flatpak` and `flatpak-builder` >= 0.8.2 to be installed on
your system. See http://flatpak.org/getting.html . For Electron apps using
Electron 6 and above, [`elfutils`](https://sourceware.org/elfutils/) is required.

Building a flatpak for an Electron app requires installing a "base" flatpak
application with electron library dependencies. That base will be autoinstalled
during an app build.

Actually running the apps will require the freedesktop runtime to be installed.
If you already are running flatpak apps you may have the freedesktop runtime
installed. The following example will install the freedesktop runtime for both
i386 and x86_64.
```
flatpak --user remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo
flatpak --user install gnome org.freedesktop.Platform/x86_64/1.6 org.freedesktop.Platform/i386/1.6
```

Apps using Electron 6 or above require a runtime version of 19.08 instead of 1.6.
Please note that as of 19.08, they have removed support for the i386 arch.

## Installation

For use from command-line:

```sh
$ npm install -g @malept/electron-installer-flatpak
```

For use in npm scripts or programmatically:

```sh
$ npm install --save-dev @malept/electron-installer-flatpak
```


## Usage

Say your Electron app lives in `path/to/app`, and has a structure like this:

```
.
├── LICENSE
├── README.md
├── node_modules
│   ├── electron-packager
│   └── electron
├── package.json
├── resources
│   ├── Icon.png
│   ├── IconTemplate.png
│   └── IconTemplate@2x.png
└── src
    ├── index.js
    ├── main
    │   └── index.js
    └── renderer
        ├── index.html
        └── index.js
```

You now run `electron-packager` to build the app for flatpak:

```
$ electron-packager . app --platform linux --arch x64 --out dist/
```

And you end up with something like this in your `dist` folder:

```
.
└── dist
    └── app-linux-x64
        ├── LICENSE
        ├── LICENSES.chromium.html
        ├── content_shell.pak
        ├── app
        ├── icudtl.dat
        ├── libgcrypt.so.11
        ├── libnode.so
        ├── locales
        ├── natives_blob.bin
        ├── resources
        ├── snapshot_blob.bin
        └── version
```

How do you turn that into a flatpak package that your users can install?

### Scripts

If you want to run `electron-installer-flatpak` through npm, install the package locally:

```
$ npm install --save-dev @malept/electron-installer-flatpak
```

Edit the `scripts` section of your `package.json`:

```json
{
  "name": "app",
  "description": "An awesome app!",
  "version": "0.0.1",
  "scripts": {
    "start": "electron .",
    "build:linux:x64": "electron-packager . --platform linux --arch x64 --out dist/",
    "build:flatpak:x64": "electron-installer-flatpak --src dist/app-linux-x64/ --dest dist/installers/ --arch x64"
  },
  "devDependencies": {
    "@malept/electron-installer-flatpak": "^0.10.1",
    "electron-packager": "^14.0.0",
    "electron": "^7.1.1"
  }
}
```

And run the script:

```
$ npm run build:flatpak:x64
```

You'll end up with the package at `dist/installer/io.atom.electron.app_master_x64.flatpak`.

### Programmatically

Install the package locally:

```
$ npm install --save-dev @malept/electron-installer-flatpak
```

An example of a script utilizing this module:

```javascript
const installer = require('@malept/electron-installer-flatpak')

const options = {
  src: 'dist/app-linux-x64/',
  dest: 'dist/installers/',
  arch: 'x64'
}

async function main (options) {
  console.log('Creating package (this may take a while)')
  try {
    await installer(options)
    console.log(`Successfully created package at ${options.dest}`)
  } catch (err) {
    console.error(err, err.stack)
    process.exit(1)
  }
}
main(options)
```

You'll end up with the package at `dist/installers/io.atom.electron.app_master_x64.flatpak`.

### Command-Line

If you want to run `@malept/electron-installer-flatpak` straight from the command-line, install the package globally:

```
$ npm install -g @malept/electron-installer-flatpak
```

And point it to your built app:

```
$ electron-installer-flatpak --src dist/app-linux-x64/ --dest dist/installers/ --arch x64
```

You'll end up with the package at `dist/installers/io.atom.electron.app_master_x64.flatpak`.

### Options

Even though you can pass most of these options through the command-line interface, it may be easier to create a configuration file:

```json
{
  "dest": "dist/installers/",
  "icon": "resources/Icon.png",
  "categories": [
    "Utility"
  ]
}
```

And pass that instead with the `config` option:

```
$ electron-installer-flatpak --src dist/app-linux-x64/ --arch x64 --config config.json
```

#### src
Type: `String`
Default: `undefined`

Path to the folder that contains your built Electron application.

#### dest
Type: `String`
Default: `undefined`

Path to the folder that will contain your flatpak installer.

#### rename
Type: `Function`
Default: `function (dest, src) { return path.join(dest, src); }`

Function that renames all files generated by the task before putting them in your `dest` folder.

#### options.id
Type: `String`
Default: the inverted hostname of `package.homepage` plus the santized `package.name`, or `io.atom.electron`

App ID of the flatpak, used in the [`id` field of a flatpak-builder manifest](https://docs.flatpak.org/en/latest/flatpak-builder-command-reference.html#flatpak-manifest) and the flatpak filename.

#### options.productName
Type: `String`
Default: `package.productName || package.name`

Name of the application (e.g. `Atom`), used in the [`Name` field of the `desktop` specification](http://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html).

#### options.genericName
Type: `String`
Default: `package.genericName || package.productName || package.name`

Generic name of the application (e.g. `Text Editor`), used in the [`GenericName` field of the `desktop` specification](http://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html).

#### options.description
Type: `String`
Default: `package.description`

Short description of the application, used in the [`Comment` field of the `desktop` specification](http://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html).

#### options.branch
Type: `String`
Default: `master`

Release branch of the flatpak, used in the [`branch` field of a flatpak-builder manifest](https://docs.flatpak.org/en/latest/flatpak-builder-command-reference.html#flatpak-manifest) and the flatpak filename, after the sanitized [app ID](#optionsid).

#### options.base
Type: `String`
Default: `io.atom.electron.BaseApp` for Electron &lt; 2, `org.electronjs.Electron2.BaseApp` for Electron >= 2

Base app to use when building the flatpak, used in the [`base` field of a flatpak-builder manifest](https://docs.flatpak.org/en/latest/flatpak-builder-command-reference.html#flatpak-manifest).

#### options.baseVersion
Type: `String`
Default: `master`

Base app version, used in the [`base-version` field of a flatpak-builder manifest](https://docs.flatpak.org/en/latest/flatpak-builder-command-reference.html#flatpak-manifest).


#### options.baseFlatpakref
Type: `String`
Default: `undefined`

Url of a flatpakref to use to auto install the base application.

#### options.runtime
Type: `String`
Default: `org.freedesktop.Platform`

Runtime id, used in the [`runtime` field of a flatpak-builder manifest](https://docs.flatpak.org/en/latest/flatpak-builder-command-reference.html#flatpak-manifest).

#### options.runtimeVersion
Type: `String`
Default: `1.6` for Electron &lt; 5, `19.08` for Electron >= 5

Runtime version, used in the [`runtime-version` field of a flatpak-builder manifest](https://docs.flatpak.org/en/latest/flatpak-builder-command-reference.html#flatpak-manifest).

#### options.sdk
Type: `String`
Default: `org.freedesktop.Sdk`

Sdk ID, used in the [`sdk` field of a flatpak-builder manifest](https://docs.flatpak.org/en/latest/flatpak-builder-command-reference.html#flatpak-manifest).

#### options.arch
Type: `String`
Default: `undefined`

Machine architecture the package is targeted to. Suggested to use node style
arches here ('ia32', 'x64'), which will be converted to flatpak style arches
('i386', 'x86_64') when calling into the actual flatpak commands. Directly
using flatpak style arches is also supported.

#### options.finishArgs
Type: `Array[String]`
Default:
```javascript
[
  // X Rendering
  '--socket=x11', '--share=ipc',
  // Open GL
  '--device=dri',
  // Audio output
  '--socket=pulseaudio',
  // Read/write home directory access
  '--filesystem=home',
  // Chromium uses a socket in tmp for its singleton check
  '--env=TMPDIR=/var/tmp',
  // Allow communication with network
  '--share=network',
  // System notifications with libnotify
  '--talk-name=org.freedesktop.Notifications'
],
```

Arguments to use when call `flatpak build-finish`, use in the [`finish-args` field of a flatpak-builder manifest](https://docs.flatpak.org/en/latest/flatpak-builder-command-reference.html#flatpak-manifest).

Changing this can be used to customize permissions of the sandbox the flatpak will run in.

#### options.files
Type: `Array[Array[source, dest]]`
Default:
```js
[],
```

Files to copy directly into the app. Should be a list of [source, dest] tuples.
Source should be a relative/absolute path to a file/directory to copy into the
flatpak, and dest should be the path inside the app install prefix (e.g.
/share/applications/)

Application assets and code will be fully handled by `electron-packager`, but
this is a useful way to install things such as appstream metadata for an app,
or dbus configuration files.

#### options.symlinks
Type: `Array[Array[target, location]]`
Default:
```js
[],
```

Symlinks to create in the app files. Should be a list of [target, location]
symlink tuples. Target can be either a relative or absolute path inside the app
install prefix, and location should be a absolute path inside the prefix to
create the symlink at.

#### options.modules
Type: `Array[Object]`
Default: `[]`

This option can be used to build extra software modules into the flatpak
application sandbox. Most electron applications will not need this, but if you
are using native node modules that require certain libraries on the system, this
may be necessary. For example, to build [libgit2](https://libgit2.github.com/)
to use with [nodegit](https://github.com/nodegit/nodegit), add the following to
your modules list.
```
modules: [
  {
    name: 'libgit2',
    cmake: true,
    configOpts: [ '-DBUILD_SHARED_LIBS:BOOL=ON', '-DTHREADSAFE=ON' ],
    sources: [{
      type: 'git',
      url: 'https://github.com/libgit2/libgit2.git',
      branch: 'maint/v0.24'
    }]
  }
]
```

See the [`modules` field of a flatpak-builder manifest](https://docs.flatpak.org/en/latest/flatpak-builder-command-reference.html#flatpak-manifest) for more details.

For apps using Electron 6 or above, the [`zypak`](https://github.com/refi64/zypak) module is
included by default so sandbox support works correctly. Since it needs to compile the module,
it only works when the host arch and target arch are the same.

#### options.bin
Type: `String`
Default: `package.name`

Relative path to the executable that will act as binary for the application, used in the [`Exec` field of the `desktop` specification](http://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html).

The generated package will contain a symlink `/usr/bin/<%= options.name %>` pointing to the path provided here.

For example, providing this configuration:

```js
{
  src: '...',
  dest: '...',
  name: 'foo',
  bin: 'resources/cli/launcher.sh'
}
```

Will create a package with the following symlink:

```
usr/bin/foo@ -> ../lib/foo/resources/cli/launcher.sh
```

And a desktop specification with the following `Exec` key:

```
Exec=foo %U
```

#### options.icon
Type: `String` or `Object[String:String]`
Default: `undefined`

Path to a single image that will act as icon for the application:

```js
{
  icon: 'resources/Icon.png'
}
```

Or multiple images with their corresponding resolutions:

```js
{
  icon: {
    '48x48': 'resources/Icon48.png',
    '64x64': 'resources/Icon64.png',
    '128x128': 'resources/Icon128.png',
    '256x256': 'resources/Icon256.png',
    'scalable': 'resources/Icon.svg'
  }
}
```
Note that the image files must be one of the types: PNG or SVG. The support for SVG works only on scalable resolution.

#### options.categories
Type: `Array[String]`
Default: `[]`

Categories in which the application should be shown in a menu, used in the [`Categories` field of the `desktop` specification](http://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html).

For possible values check out the [Desktop Menu Specification](http://standards.freedesktop.org/menu-spec/latest/apa.html).

#### options.mimeType
Type: `Array[String]`
Default: `[]`

MIME types the application is able to open, used in the [`MimeType` field of the `desktop` specification](http://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html).


## Meta

* Code: `git clone https://github.com/malept/electron-installer-flatpak.git`
* Home: <https://github.com/malept/electron-installer-flatpak/>


## Contributors

* Daniel Perez Alvarez ([unindented@gmail.com](mailto:unindented@gmail.com))
* Matt Watson ([mattdangerw@gmail.com](mailto:mattdangerw@gmail.com))
* [Mark Lee](https://github.com/malept)

## License

Copyright (c) 2016 Daniel Perez Alvarez ([unindented.org](https://unindented.org/)). This is free software, and may be redistributed under the terms specified in the LICENSE file.


NEWS.md
# `@malept/electron-installer-flatpak` - Changes by Version

## [Unreleased]

[Unreleased]: https://github.com/malept/electron-installer-flatpak/compare/v0.11.4...master

## [0.11.4] - 2021-07-14

[0.11.4]: https://github.com/malept/electron-installer-flatpak/compare/v0.11.3...v0.11.4

### Fixed

* typo in the CLI entry in `package.json` (#43)

## [0.11.3] - 2021-05-09

[0.11.3]: https://github.com/malept/electron-installer-flatpak/compare/v0.11.2...v0.11.3

### Fixed

* Update zypak to v2021.02 (malept#40)

## [0.11.2] - 2020-06-06

[0.11.2]: https://github.com/malept/electron-installer-flatpak/compare/v0.11.1...v0.11.2

### Fixed

* Provide defaults for the `*Flatpakref` options (malept#26)

## [0.11.1] - 2020-01-21

[0.11.1]: https://github.com/malept/electron-installer-flatpak/compare/v0.11.0...v0.11.1

### Changed

* Upgraded `electron-installer-common` to `^0.10.0`

## [0.11.0] - 2019-12-17

[0.11.0]: https://github.com/malept/electron-installer-flatpak/compare/v0.10.1...v0.11.0

### Added

* Support for zypak for Electron >= 5 (malept#16)

### Fixed

* Force baseVersion and runtimeVersion to strings (port of endlessm#27)
* Don't provide full access to `/tmp` (port of endlessm#27)

### Removed

* Node &lt; 8 support (malept#18)

## [0.10.1] - 2019-10-29

[0.10.1]: https://github.com/malept/electron-installer-flatpak/compare/v0.10.0...v0.10.1

### Fixed

* Throw a better error when a homepage URL cannot be parsed (malept#14)

## [0.10.0] - 2019-06-15

[0.10.0]: https://github.com/malept/electron-installer-flatpak/compare/v0.9.0...v0.10.0

### Added

* Support for Electron >= 2.0 (malept#6)

### Removed

* Node &lt; 8 support (malept#5)

## [0.9.0] - 2019-03-13

[0.9.0]: https://github.com/malept/electron-installer-flatpak/compare/v0.8.0...v0.9.0

### Added

* Support for custom `flatpak-builder` args (malept#4)

### Fixed

* Append `--assumeyes` to the `flatpak-builder` arguments when `flatpak-builder` is new enough, to
  fix a bug with auto-installing flatpak dependencies (malept#4)

### Removed

* Node &lt; 6 support (malept#1)

----

For versions prior to 0.9.0, please see `git log`.


test/cli.js
'use strict'

const assert = require('assert')
const fs = require('fs-extra')
const { spawn } = require('electron-installer-common')

describe('cli', function () {
  this.timeout(90000)

  describe('with an app with asar', function () {
    const dest = 'test/fixtures/out/foo/'
    const expectedFlatpak = `${dest}org.unindented.footest_stable_x64.flatpak`

    before(async () => spawn('./bin/electron-installer-flatpak.js', [
      '--src', 'test/fixtures/app-with-asar/',
      '--dest', dest,
      '--arch', 'x64'
    ]))

    after(async () => fs.remove(dest))

    it('generates a `.flatpak` package', async () => assert.ok(await fs.pathExists(expectedFlatpak), `${expectedFlatpak} not created`))
  })

  describe('with an app without asar', function () {
    const dest = 'test/fixtures/out/bar/'
    const expectedFlatpak = `${dest}com.foo.bartest_stable_x64.flatpak`

    before(async () => spawn('./bin/electron-installer-flatpak.js', [
      '--src', 'test/fixtures/app-without-asar/',
      '--dest', dest,
      '--arch', 'x64'
    ]))

    it('generates a `.flatpak` package', async () => assert.ok(await fs.pathExists(expectedFlatpak), `${expectedFlatpak} not created`))
  })
})


test/ci/test.sh
#!/bin/bash -e

if [[ -z "$USE_DOCKER" ]]; then
    npm test
else
    sudo docker run --privileged --interactive --tty --volume $(pwd):/code malept/electron-forge-container:latest /bin/bash -c "cd /code &&
        /code/test/ci/install_runtimes.sh &&
        DEBUG=$DEBUG npm test"
fi


test/ci/script.sh
#!/bin/bash -e

if [[ "$USE_DOCKER" = "true" ]]; then
    sudo docker run --privileged --interactive --tty --volume $(pwd):/code malept/electron-forge-container:latest /bin/bash -c "cd /code &&
        mkdir -p /var/run/dbus &&
        dbus-daemon --system &&
        /code/test/ci/install_runtimes.sh &&
        DEBUG=$DEBUG npm test"
else
    npm test
fi


src/cli.js
'use strict'

const { omit } = require('lodash')
const yargs = require('yargs')

const installer = require('./installer')
const { description, version } = require('../package.json')

module.exports = {
  parseArgs: function parseArgs (args) {
    return yargs
      .version(version)
      .usage(`${description}\n\nUsage: $0 --src <inputdir> --dest <outputdir> --arch <architecture> [...]`)
      .option('src', {
        describe: 'Directory that contains your built Electron app (e.g. with `electron-packager`)',
        demand: true
      })
      .option('dest', {
        describe: 'Directory that will contain the resulting Flatpak',
        demand: true
      })
      .option('arch', {
        describe: 'Target machine architecture for the package',
        demand: true
      })
      .option('config', {
        describe: 'JSON file that contains the metadata for your application',
        config: true
      })
      .example('$0 --src dist/app/ --dest dist/installer/ --arch ia32', 'use metadata from `dist/app/`')
      .example('$0 --src dist/app/ --dest dist/installer/ --config config.json', 'use metadata from `config.json`')
      .wrap(null)
      .parse(args)
  },
  run: async function run (args) {
    const argv = module.exports.parseArgs()

    console.log('Creating package (this may take a while)')

    const options = omit(argv, ['$0', '_', 'version'])
    await installer(options)
    console.log(`Successfully created package at ${argv.dest}`)
  }
}


src/getappid.js
'use strict'

const url = require('url')

function sanitizePackageNameParts (parts) {
  return parts.map(part => part.replace(/[^a-z0-9]/gi, '_').replace(/^[0-9]/, '_$&'))
}

function parseHomepageURL (website) {
  try {
    return new url.URL(website)
  } catch (e) {
    throw new Error(`Could not parse the homepage from package.json as a URL:\n${e.message}`)
  }
}

module.exports = {
  getAppID: function getAppID (name, website) {
    let host = 'electron.atom.io'
    if (website) {
      const urlObject = parseHomepageURL(website)
      if (urlObject.host) {
        host = urlObject.host
      }
    }
    let parts = host.split('.')
    if (parts[0] === 'www') {
      parts.shift()
    }
    parts = sanitizePackageNameParts(parts.reverse())
    parts.push(name)
    let appID = parts.join('.')
    while (appID.length > 255) {
      parts.unshift()
      appID = parts.join('.')
    }
    return appID
  }
}


example/src/main/index.js
'use strict'

const path = require('path')
const { app, BrowserWindow, Menu, Tray } = require('electron')

let quitting = false
let tray = null
let win = null

function createMenu () {
  const appMenu = Menu.buildFromTemplate([
    {
      label: 'File',
      submenu: [
        {
          label: 'Quit',
          accelerator: 'CmdOrCtrl+Q',
          click: () => {
            app.quit()
          }
        }
      ]
    }
  ])
  Menu.setApplicationMenu(appMenu)
}

function createTray () {
  const variant = (process.platform === 'darwin' ? 'Black' : 'White')

  tray = new Tray(path.resolve(__dirname, `../../resources/Icon${variant}Template.png`))

  const trayMenu = Menu.buildFromTemplate([
    {
      label: 'Preferences...',
      click: () => {
        win.show()
      }
    },
    {
      type: 'separator'
    },
    {
      label: 'Quit',
      click: () => {
        app.quit()
      }
    }
  ])
  tray.setContextMenu(trayMenu)
}

function createWindow () {
  win = new BrowserWindow({
    width: 800,
    height: 600,
    show: false,
    icon: path.resolve(__dirname, '../../resources/Icon.png')
  })
  win.loadFile(path.resolve(__dirname, '..', 'renderer', 'index.html'))

  win.on('close', (evt) => {
    if (quitting) {
      return
    }

    evt.preventDefault()
    win.hide()
  })

  win.on('closed', () => {
    tray = null
    win = null
  })
}

app.on('before-quit', () => {
  quitting = true
})

app.on('window-all-closed', () => {
  app.quit()
})

app.on('ready', () => {
  createMenu()
  createTray()
  createWindow()
})


src/installer.js
'use strict'

const _ = require('lodash')
const common = require('electron-installer-common')
const debug = require('debug')
const flatpak = require('@malept/flatpak-bundler')
const { getAppID } = require('./getappid')
const path = require('path')
const semver = require('semver')

const defaultLogger = debug('electron-installer-flatpak')
const defaultRename = (dest, src) => path.join(dest, src)

class FlatpakInstaller extends common.ElectronInstaller {
  get appIdentifier () {
    return this.options.id
  }

  get baseAppDir () {
    return ''
  }

  get contentFunctions () {
    return [
      'copyApplication',
      'copyLinuxIcons',
      'createBinWrapper',
      'createCopyright',
      'createDesktopFile'
    ]
  }

  get defaultDesktopTemplatePath () {
    return path.join(this.resourcesDir, 'desktop.ejs')
  }

  get flatpakrefs () {
    const arch = flatpak.translateArch(this.options.arch)
    return {
      baseFlatpakref: this.options.baseFlatpakref || `app/${this.options.base}/${arch}/${this.options.baseVersion}`,
      runtimeFlatpakref: this.options.runtimeFlatpakref || `runtime/${this.options.runtime}/${arch}/${this.options.runtimeVersion}`,
      sdkFlatpakref: this.options.sdkFlatpakref || `runtime/${this.options.sdk}/${arch}/${this.options.runtimeVersion}`
    }
  }

  get resourcesDir () {
    return path.resolve(__dirname, '../resources')
  }

  async createBinWrapper () {
    if (await this.requiresSandboxWrapper()) {
      await this.createTemplatedFile(path.join(this.resourcesDir, 'electron-wrapper.ejs'), path.join(this.stagingDir, this.baseAppDir, 'bin', 'electron-wrapper'), 0o755)
    }
  }

  async createDesktopFile () {
    this.options.desktopExec = (await this.requiresSandboxWrapper()) ? 'electron-wrapper' : this.options.bin
    await super.createDesktopFile()
  }

  async determineBaseRuntimeAndSDK () {
    const baseConfig = {
      branch: 'stable',
      baseVersion: 'stable',
      runtime: 'org.freedesktop.Platform',
      sdk: 'org.freedesktop.Sdk'
    }
    if (await this.requiresSandboxWrapper()) {
      baseConfig.runtimeVersion = '19.08'
    } else {
      baseConfig.runtimeVersion = '1.6'
    }
    if (semver.gte(await common.readElectronVersion(this.userSupplied.src), '2.0.0-beta.1')) {
      baseConfig.base = 'org.electronjs.Electron2.BaseApp'
    } else {
      baseConfig.base = 'io.atom.electron.BaseApp'
    }

    return baseConfig
  }

  /**
   * Generate the hash of default options for the installer. Some come from the info
   * read from `package.json`, and some are hardcoded.
   */
  async generateDefaults () {
    const pkg = (await common.readMetadata(this.userSupplied)) || {}
    const modules = []
    if (await this.requiresSandboxWrapper()) {
      modules.push({
        name: 'zypak',
        sources: [
          {
            type: 'git',
            url: 'https://github.com/refi64/zypak',
            tag: 'v2021.02'
          }
        ]
      })
    }
    this.defaults = {
      ...common.getDefaultsFromPackageJSON(pkg),
      ...(await this.determineBaseRuntimeAndSDK()),
      id: getAppID(pkg.name, common.getHomePage(pkg)),
      extraFlatpakBuilderArgs: [],
      finishArgs: [
        // X Rendering
        '--socket=x11', '--share=ipc',
        // OpenGL
        '--device=dri',
        // Audio output
        '--socket=pulseaudio',
        // Read/write home directory access
        '--filesystem=home',
        // Chromium uses a socket in tmp for its singleton check
        '--env=TMPDIR=/var/tmp',
        // Allow communication with network
        '--share=network',
        // System notifications with libnotify
        '--talk-name=org.freedesktop.Notifications'
      ],
      modules,

      icon: path.resolve(__dirname, '../resources/icon.png'),
      files: [],
      symlinks: []
    }

    return this.defaults
  }

  async requiresSandboxWrapper () {
    if (typeof this._requiresSandboxWrapper === 'undefined') {
      this._requiresSandboxWrapper = await common.hasSandboxHelper(this.sourceDir)
    }

    return this._requiresSandboxWrapper
  }

  /**
   * Bundle everything using `flatpak-bundler`.
   */
  async createBundle () {
    const name = `${this.appIdentifier}_${this.options.branch}_${this.options.arch}.flatpak`
    const dest = this.options.rename(this.options.dest, name)
    this.options.logger(`Creating package at ${dest}`)
    const extraExports = []
    if (this.options.icon && !_.isObject(this.options.icon)) {
      extraExports.push(this.pixmapIconPath)
    }

    const files = [
      [this.stagingDir, '/']
    ]
    const symlinks = [
      [path.join('/lib', this.appIdentifier, this.options.bin), path.join('/bin', this.options.bin)]
    ]

    const command = (await this.requiresSandboxWrapper()) ? 'electron-wrapper' : this.options.bin

    return flatpak.bundle({
      id: this.options.id,
      branch: this.options.branch,
      base: this.options.base,
      baseVersion: this.options.baseVersion.toString(),
      runtime: this.options.runtime,
      runtimeVersion: this.options.runtimeVersion.toString(),
      sdk: this.options.sdk,
      finishArgs: this.options.finishArgs,
      command,
      modules: this.options.modules
    }, {
      ...this.flatpakrefs,
      arch: this.options.arch,
      bundlePath: dest,
      extraExports: extraExports,
      extraFlatpakBuilderArgs: this.options.extraFlatpakBuilderArgs,
      files: files.concat(this.options.files),
      symlinks: symlinks.concat(this.options.symlinks)
    })
  }
}

/* ************************************************************************** */

module.exports = async data => {
  data.rename = data.rename || defaultRename
  data.logger = data.logger || defaultLogger

  const installer = new FlatpakInstaller(data)

  await installer.generateDefaults()
  await installer.generateOptions()
  data.logger(`Creating package with options\n${JSON.stringify(installer.options, null, 2)}`)
  await installer.createStagingDir()
  await installer.createContents()
  await installer.createBundle()
  data.logger(`Successfully created package at ${installer.options.dest}`)
  return installer.options
}

module.exports.Installer = FlatpakInstaller
